src/bitvec.rs

```diff
@@ -0,0 +1,371 @@
+//! longer a dense bitvector; the chunked adaptive-sparse data
+//! structure here has better performance.
+
+use fxhash::FxHashMap;
+use std::cell::Cell;
+
+const SMALL_ELEMS: usize = 12;
+
+/// A hybrid large/small-mode sparse mapping from integer indices to
+/// elements.
+///
+/// The trailing `(u32, u64)` elements in each variant is a one-item
+/// cache to allow fast access when streaming through.
+#[derive(Clone, Debug)]
+enum AdaptiveMap {
+    Small {
+        len: u32,
+        keys: [u32; SMALL_ELEMS],
+        values: [u64; SMALL_ELEMS],
+    },
+    Large(FxHashMap<u32, u64>),
+}
+
+const INVALID: u32 = 0xffff_ffff;
+
+impl AdaptiveMap {
+    fn new() -> Self {
+        Self::Small {
+            len: 0,
+            keys: [INVALID; SMALL_ELEMS],
+            values: [0; SMALL_ELEMS],
+        }
+    }
+
+    /// Expand into `Large` mode if we are at capacity and have no
+    /// zero-value pairs that can be trimmed.
+    #[inline(never)]
+    fn expand(&mut self) {
+        match self {
+            &mut Self::Small {
+                ref mut len,
+                ref mut keys,
+                ref mut values,
+            } => {
+                // Note: we *may* remain as `Small` if there are any
+                // zero elements. Try removing them first, before we
+                // commit to a memory allocation.
+                if values.iter().any(|v| *v == 0) {
+                    let mut out = 0;
+                    for i in 0..(*len as usize) {
+                        if values[i] == 0 {
+                            continue;
+                        }
+                        if out < i {
+                            keys[out] = keys[i];
+                            values[out] = values[i];
+                        }
+                        out += 1;
+                    }
+                    *len = out as u32;
+                } else {
+                    let mut map = FxHashMap::default();
+                    for i in 0..(*len as usize) {
+                        map.insert(keys[i], values[i]);
+                    }
+                    *self = Self::Large(map);
+                }
+            }
+            _ => {}
+        }
+    }
+    #[inline(always)]
+    fn get_or_insert<'a>(&'a mut self, key: u32) -> &'a mut u64 {
+        // Check whether the key is present and we are in small mode;
+        // if no to both, we need to expand first.
+        let (needs_expand, small_mode_idx) = match self {
+            &mut Self::Small { len, ref keys, .. } => {
+                // Perform this scan but do not return right away;
+                // doing so runs into overlapping-borrow issues
+                // because the current non-lexical lifetimes
+                // implementation is not able to see that the `self`
+                // mutable borrow on return is only on the
+                // early-return path.
+                let small_mode_idx = keys.iter().position(|k| *k == key);
```

Comment:
```suggestion
                let small_mode_idx = keys.iter().take(len as usize).position(|k| *k == key);
```

Reply:
Ah, a silly bug on my part; thank you for finding this!